src/backend/distributed/commands/table.c

```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
 {
-	List *commandList = alterTableStatement->cmds;
+	bool rightHasPostgresLocalTable = false;
+	bool rightHasNonDistTable = false;
 
-	LOCKMODE lockmode = AlterTableGetLockLevel(commandList);
-	Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);
+	Oid rightRelationId = InvalidOid;
+	List *rightRelationIdList =
+		GetAlterTableAddFKeyRightRelationIdList(alterTableStatement);
+	foreach_oid(rightRelationId, rightRelationIdList)
+	{
+		if (!IsCitusTable(rightRelationId))
+		{
+			rightHasPostgresLocalTable = true;
+		}
+		else if (IsCitusTableType(rightRelationId, CITUS_TABLE_WITH_NO_DIST_KEY))
+		{
+			rightHasNonDistTable = true;
+		}
+	}
 
-	if (IsCitusTable(leftRelationId))
+	LOCKMODE lockmode = AlterTableGetLockLevel(alterTableStatement->cmds);
+	Oid leftRelationId = AlterTableLookupRelation(alterTableStatement, lockmode);
+	bool leftIsPostgresTable = !IsCitusTable(leftRelationId);
+	if (leftIsPostgresTable)
 	{
-		/* left relation is not a postgres local table, */
-		return;
+		return rightHasNonDistTable;
 	}
 
-	List *alterTableFKeyConstraints =
-		GetAlterTableStmtFKeyConstraintList(alterTableStatement);
-	Constraint *constraint = NULL;
-	foreach_ptr(constraint, alterTableFKeyConstraints)
+	bool leftIsNonDistTable = IsCitusTableType(leftRelationId,
+											   CITUS_TABLE_WITH_NO_DIST_KEY);
+	if (leftIsNonDistTable)
 	{
-		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
-											   alterTableStatement->missing_ok);
-		if (IsCitusTableType(rightRelationId, CITUS_LOCAL_TABLE))
+		return rightHasPostgresLocalTable;
+	}
+
+	return false;
+}
+
+
+/*
+ * ConvertPostgresLocalTablesToCitusLocalTables converts each postgres table
+ * involved in foreign keys to be defined by given alter table command and the
+ * other tables connected to them via a foreign key graph to citus local tables.
+ */
+static void
+ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	RangeVar *leftRelationRangeVar = alterTableStatement->relation;
+	List *relationRangeVarList = lappend(rightRelationRangeVarList, leftRelationRangeVar);
+
+	/*
+	 * To prevent deadlocks, sort the list before converting each postgres local
+	 * table to a citus local table.
+	 */
+	relationRangeVarList = SortList(relationRangeVarList, CompareRangeVarsByOid);
+
+	/*
+	 * Here we should operate on RangeVar objects since relations oid's would
+	 * change in below loop due to CreateCitusLocalTable.
+	 */
+	RangeVar *relationRangeVar;
+	foreach_ptr(relationRangeVar, relationRangeVarList)
+	{
+		List *commandList = alterTableStatement->cmds;
+		LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+		bool missingOk = alterTableStatement->missing_ok;
+		Oid rightRelationId = RangeVarGetRelid(relationRangeVar, lockMode, missingOk);
```

Comment:
minor: we should probably not use the name `rightRelationId ` as the left relation is also on the list

Reply:
f242d116e61aa05d66421bae5ee667c62bf4f5f2